### PR TITLE
feat: user can cancel python execution

### PR DESCRIPTION
**Pull request by ddimaria**
CleanShot 2023-11-17 at 13 33 45@2x

- [x] Show Python running indicator in bottom bar
- [x] Track kill Python event in Mixpanel
- [x] Add stop button to python code editor
- [x] When stoping python, complete the compute cycle, finalize the transaction, and send message to the editor
- [x] Don't error on dependent code cells
- [x] Show loading indicator for reloading pyodide

### CODE REVIEW

File: src/gridGL/interaction/keyboard/keyboardViewport.ts [hunk 1]
```diff
@@ -34,7 +35,8 @@ export function keyboardViewport(options: {
     return true;
   }
 
-  if (event.key === 'Escape') {
+  if (!(event.metaKey || event.ctrlKey) && event.key === 'Escape') {
+    console.log('escape presentation');
```
Comment (davidfig) [inline src/gridGL/interaction/keyboard/keyboardViewport.ts:39]: console.log

**Review by davidfig — APPROVED** [review]

**Comment by davidfig** [comment]
Can you add a rust test with a dependent cell where you start a python execution, cancel it, and the dependent cell is not triggered?

File: src/ui/menus/CodeEditor/CodeEditorHeader.tsx [hunk 2]
```diff
@@ -101,6 +102,21 @@ export const CodeEditorHeader = (props: Props) => {
             <HelpOutline fontSize="small" />
           </IconButton>
         </TooltipHint>
+        {hasPermission && (
+          <TooltipHint title="Cancel execution" shortcut={`${KeyboardSymbols.Command}␛`} placement="bottom">
+            <span>
+              <IconButton
+                id="QuadraticCodeEditorCalcelButtonID"
```
Comment (davidkircos) [inline src/ui/menus/CodeEditor/CodeEditorHeader.tsx:109]: is this ID Used anywhere?

Comment (ddimaria) [reply src/ui/menus/CodeEditor/CodeEditorHeader.tsx:109]: No, and the spelling is off.  I added it b/c I though it was used in testing (or elsewhere).  Feel free to remove.

File: src/ui/menus/CodeEditor/CodeEditorHeader.tsx [hunk 3]
```diff
@@ -101,6 +102,21 @@ export const CodeEditorHeader = (props: Props) => {
             <HelpOutline fontSize="small" />
           </IconButton>
         </TooltipHint>
+        {hasPermission && (
+          <TooltipHint title="Cancel execution" shortcut={`${KeyboardSymbols.Command}␛`} placement="bottom">
```
Comment (jimniels) [inline src/ui/menus/CodeEditor/CodeEditorHeader.tsx:106]: That `ESC` symbol is _barely_ legible.

CleanShot 2023-11-17 at 12 58 08@2x

We should just spell it out with normal characters ([like we do here for close](https://github.com/quadratichq/quadratic/pull/823/files#diff-f9f7c587c87bda509d9809bc92299954c654ff3077f93ae039d00d22735d3d17R135))

CleanShot 2023-11-17 at 12 58 13@2x

In fact, can we change `ESC` to be `Esc` in both places? e.g. `⌘ Esc`

File: src/gridGL/interaction/keyboard/keyboardViewport.ts [hunk 4]
```diff
@@ -105,5 +107,10 @@ export function keyboardViewport(options: {
     return true;
   }
 
+  // Command + Escape
+  if ((event.metaKey || event.ctrlKey) && event.key === 'Escape') {
+    pythonWebWorker.restartFromUser();
+  }
```
Comment (jimniels) [inline src/gridGL/interaction/keyboard/keyboardViewport.ts:113]: Seems like we don't want the user to be able to trigger this if there's no actual code running...yeah? I can hit `Cmd + Esc` while nothing is executing and it reloads python.

CleanShot 2023-11-17 at 13 05 45

File: src/ui/menus/CodeEditor/CodeEditorHeader.tsx (hunk 3 above)
Comment (jimniels) [reply src/ui/menus/CodeEditor/CodeEditorHeader.tsx:106]: Ha apparently `⎋` is the symbol for escape — I did not know that?

`⌘⎋` seems a bit too obscure IMO, it's not even printed on most keyboards

File: src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 5]
```diff
@@ -138,10 +145,10 @@ export const CodeEditor = () => {
...
   const onKeyDownEditor = (event: React.KeyboardEvent<HTMLDivElement>) => {
     // Esc
-    if (!(event.metaKey || event.ctrlKey) && event.key === 'Escape') {
-      event.preventDefault();
-      closeEditor(false);
-    }
+    // if (!(event.metaKey || event.ctrlKey) && event.key === 'Escape') {
+    //   event.preventDefault();
+    //   closeEditor(false);
+    // }
```
Comment (jimniels) [inline src/ui/menus/CodeEditor/CodeEditor.tsx:151]: We should just remove these, yeah?

**Review by jimniels — DISMISSED** [review]

File: src/ui/menus/CodeEditor/CodeEditor.tsx (hunk 5 above)
Comment (ddimaria) [reply src/ui/menus/CodeEditor/CodeEditor.tsx:151]: There was a recent merge from another branch, I'll need to go through all of the escapes later today.

**Comment by jimniels** [comment]
I updated the design around python execution status in both the code editor header and the bottom bar.

Now in the bottom bar, you can access the state of python via a dropdown. It gives you the ability to cancel execution for long running transactions.

CleanShot 2023-11-17 at 15 35 02@2x

CleanShot 2023-11-17 at 15 35 11@2x

CleanShot 2023-11-17 at 15 35 28@2x

CleanShot 2023-11-17 at 15 35 46@2x


It also replaces the previous toast for "Transaction busy" and now the dropdown pops up automatically if you try to edit a cell while python is executing

CleanShot 2023-11-17 at 15 35 19@2x

**Pull request merged**